### PR TITLE
[TraceQL] Remove unnecessary spanset layer, iterators, and predicates

### PR DESCRIPTION
**Pull request by mdisibio**
**What this PR does**:
The changes are:
1. Remove the intermediate spansets in batchCollector. Now it updates the spans in-place on the way to the traceCollector, which still creates the final spanset.
2. Remove unnecessary OrPredicates and UnionIterators.
3. Various other cleanups and simplifications

Benchmarks
```
name                                              old time/op    new time/op    delta
BackendBlockTraceQL/spanAttNameNoMatch              3.94ms ± 2%    3.91ms ± 1%     ~     (p=0.222 n=5+5)
BackendBlockTraceQL/spanAttValNoMatch               35.2ms ± 1%    34.4ms ± 0%   -2.27%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttValMatch                 39.4ms ± 1%    37.4ms ± 2%   -5.14%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicNoMatch         3.84ms ± 1%    3.77ms ± 7%     ~     (p=0.222 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicMatch           70.3ms ± 1%    63.3ms ± 1%  -10.01%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicRegexNoMatch    4.30ms ± 1%    4.26ms ± 2%     ~     (p=0.421 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicRegexMatch       107ms ± 2%      98ms ± 0%   -8.11%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttNameNoMatch          3.17ms ± 2%    3.15ms ± 2%     ~     (p=0.548 n=5+5)
BackendBlockTraceQL/resourceAttValNoMatch           3.00ms ± 1%    3.02ms ± 2%     ~     (p=1.000 n=5+5)
BackendBlockTraceQL/resourceAttValMatch              625ms ± 4%     591ms ± 3%   -5.39%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch     3.02ms ± 0%    3.02ms ± 1%     ~     (p=0.421 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicMatch       66.5ms ± 1%    66.4ms ± 4%     ~     (p=0.222 n=5+5)
BackendBlockTraceQL/mixedNameNoMatch                 699ms ± 0%     702ms ± 1%     ~     (p=0.548 n=5+5)
BackendBlockTraceQL/mixedValNoMatch                  673ms ± 2%     670ms ± 1%     ~     (p=0.548 n=5+5)
BackendBlockTraceQL/mixedValMixedMatchAnd           3.15ms ± 1%    3.14ms ± 1%     ~     (p=0.548 n=5+5)
BackendBlockTraceQL/mixedValMixedMatchOr             574ms ± 1%     564ms ± 1%   -1.82%  (p=0.016 n=5+5)
BackendBlockTraceQL/mixedValBothMatch               3.01ms ± 1%    3.03ms ± 1%     ~     (p=0.548 n=5+5)
BackendBlockTraceQL/count                            2.76s ± 1%     2.48s ± 2%  -10.27%  (p=0.008 n=5+5)
BackendBlockTraceQL/desc                             3.38s ± 1%     3.32s ± 2%   -1.83%  (p=0.032 n=5+5)

name                                              old speed      new speed      delta
BackendBlockTraceQL/spanAttNameNoMatch             814MB/s ± 2%   820MB/s ± 1%     ~     (p=0.222 n=5+5)
BackendBlockTraceQL/spanAttValNoMatch              219MB/s ± 1%   224MB/s ± 0%   +2.32%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttValMatch               94.7MB/s ± 1%  99.8MB/s ± 2%   +5.43%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicNoMatch        762MB/s ± 1%   776MB/s ± 7%     ~     (p=0.222 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicMatch          317MB/s ± 1%   352MB/s ± 1%  +11.13%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicRegexNoMatch   680MB/s ± 1%   687MB/s ± 2%     ~     (p=0.421 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicRegexMatch     208MB/s ± 2%   227MB/s ± 0%   +8.81%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttNameNoMatch         399MB/s ± 2%   401MB/s ± 2%     ~     (p=0.548 n=5+5)
BackendBlockTraceQL/resourceAttValNoMatch          177MB/s ± 1%   177MB/s ± 2%     ~     (p=1.000 n=5+5)
BackendBlockTraceQL/resourceAttValMatch           62.4MB/s ± 4%  66.0MB/s ± 3%   +5.68%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch    222MB/s ± 0%   223MB/s ± 1%     ~     (p=0.421 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicMatch      548MB/s ± 1%   550MB/s ± 3%     ~     (p=0.222 n=5+5)
BackendBlockTraceQL/mixedNameNoMatch              18.7MB/s ± 0%  18.6MB/s ± 1%     ~     (p=0.476 n=5+5)
BackendBlockTraceQL/mixedValNoMatch               19.4MB/s ± 2%  19.5MB/s ± 1%     ~     (p=0.452 n=5+5)
BackendBlockTraceQL/mixedValMixedMatchAnd          400MB/s ± 1%   402MB/s ± 1%     ~     (p=0.548 n=5+5)
BackendBlockTraceQL/mixedValMixedMatchOr          44.4MB/s ± 1%  45.3MB/s ± 1%   +1.86%  (p=0.016 n=5+5)
BackendBlockTraceQL/mixedValBothMatch              223MB/s ± 1%   222MB/s ± 1%     ~     (p=0.548 n=5+5)
BackendBlockTraceQL/count                         12.9MB/s ± 1%  14.4MB/s ± 2%  +11.43%  (p=0.008 n=5+5)
BackendBlockTraceQL/desc                          14.1MB/s ± 1%  14.4MB/s ± 2%   +1.89%  (p=0.032 n=5+5)

name                                              old MB_io/op   new MB_io/op   delta
BackendBlockTraceQL/spanAttNameNoMatch                3.21 ± 0%      3.21 ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttValNoMatch                 7.71 ± 0%      7.71 ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttValMatch                   3.73 ± 0%      3.73 ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicNoMatch           2.93 ± 0%      2.93 ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicMatch             22.3 ± 0%      22.3 ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicRegexNoMatch      2.93 ± 0%      2.93 ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicRegexMatch        22.3 ± 0%      22.3 ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttNameNoMatch            1.26 ± 0%      1.26 ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttValNoMatch             0.53 ± 0%      0.53 ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttValMatch               39.0 ± 0%      39.0 ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch       0.67 ± 0%      0.67 ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttIntrinsicMatch         36.5 ± 0%      36.5 ± 0%     ~     (all equal)
BackendBlockTraceQL/mixedNameNoMatch                  13.1 ± 0%      13.1 ± 0%     ~     (all equal)
BackendBlockTraceQL/mixedValNoMatch                   13.1 ± 0%      13.1 ± 0%     ~     (all equal)
BackendBlockTraceQL/mixedValMixedMatchAnd             1.26 ± 0%      1.26 ± 0%     ~     (all equal)
BackendBlockTraceQL/mixedValMixedMatchOr              25.5 ± 0%      25.5 ± 0%     ~     (all equal)
BackendBlockTraceQL/mixedValBothMatch                 0.67 ± 0%      0.67 ± 0%     ~     (all equal)
BackendBlockTraceQL/count                             35.7 ± 0%      35.7 ± 0%     ~     (all equal)
BackendBlockTraceQL/desc                              47.7 ± 0%      47.7 ± 0%     ~     (all equal)

name                                              old alloc/op   new alloc/op   delta
BackendBlockTraceQL/spanAttNameNoMatch              1.37MB ± 0%    1.37MB ± 0%   +0.00%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttValNoMatch               92.2MB ± 1%    91.5MB ± 1%     ~     (p=0.222 n=5+5)
BackendBlockTraceQL/spanAttValMatch                 1.38MB ± 0%    1.38MB ± 0%     ~     (p=0.548 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicNoMatch         1.59MB ± 0%    1.59MB ± 0%   +0.00%  (p=0.032 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicMatch           3.83MB ± 0%    3.86MB ± 0%   +0.59%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicRegexNoMatch    1.80MB ± 0%    1.80MB ± 0%   +0.00%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicRegexMatch      7.94MB ± 0%    7.96MB ± 0%   +0.26%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttNameNoMatch          1.27MB ± 0%    1.27MB ± 0%   +0.00%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttValNoMatch           1.27MB ± 0%    1.27MB ± 0%   +0.00%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttValMatch              184MB ± 0%     182MB ± 1%   -0.77%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch     1.27MB ± 0%    1.27MB ± 0%   +0.00%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicMatch       2.19MB ± 0%    2.19MB ± 0%   -0.38%  (p=0.008 n=5+5)
BackendBlockTraceQL/mixedNameNoMatch                 101MB ± 3%     100MB ± 3%     ~     (p=0.087 n=5+5)
BackendBlockTraceQL/mixedValNoMatch                  109MB ± 0%     109MB ± 1%     ~     (p=0.254 n=5+4)
BackendBlockTraceQL/mixedValMixedMatchAnd           1.27MB ± 0%    1.27MB ± 0%     ~     (p=0.095 n=5+5)
BackendBlockTraceQL/mixedValMixedMatchOr            3.95MB ± 0%    3.96MB ± 0%     ~     (p=0.063 n=5+4)
BackendBlockTraceQL/mixedValBothMatch               1.27MB ± 0%    1.27MB ± 0%   +0.00%  (p=0.008 n=5+5)
BackendBlockTraceQL/count                            937MB ± 0%     898MB ± 1%   -4.16%  (p=0.008 n=5+5)
BackendBlockTraceQL/desc                             690MB ± 0%     674MB ± 0%   -2.32%  (p=0.008 n=5+5)

name                                              old allocs/op  new allocs/op  delta
BackendBlockTraceQL/spanAttNameNoMatch               18.7k ± 0%     18.7k ± 0%     ~     (p=1.000 n=5+5)
BackendBlockTraceQL/spanAttValNoMatch                19.0k ± 0%     19.0k ± 0%     ~     (p=1.000 n=5+5)
BackendBlockTraceQL/spanAttValMatch                  19.1k ± 0%     19.1k ± 0%     ~     (p=1.000 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicNoMatch          18.7k ± 0%     18.7k ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicMatch            30.1k ± 0%     30.1k ± 0%   -0.30%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicRegexNoMatch     21.0k ± 0%     21.0k ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicRegexMatch       73.0k ± 0%     72.9k ± 0%   -0.13%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttNameNoMatch           18.7k ± 0%     18.7k ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttValNoMatch            18.7k ± 0%     18.7k ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttValMatch              1.69M ± 0%     1.68M ± 0%   -0.76%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch      18.7k ± 0%     18.7k ± 0%   +0.01%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicMatch        24.0k ± 0%     24.0k ± 0%   -0.10%  (p=0.008 n=5+5)
BackendBlockTraceQL/mixedNameNoMatch                 21.1k ± 0%     21.0k ± 0%   -0.07%  (p=0.008 n=5+5)
BackendBlockTraceQL/mixedValNoMatch                  21.9k ± 0%     21.9k ± 0%   -0.09%  (p=0.016 n=5+4)
BackendBlockTraceQL/mixedValMixedMatchAnd            18.7k ± 0%     18.7k ± 0%   -0.01%  (p=0.008 n=5+5)
BackendBlockTraceQL/mixedValMixedMatchOr             31.2k ± 0%     31.1k ± 0%   -0.35%  (p=0.008 n=5+5)
BackendBlockTraceQL/mixedValBothMatch                18.7k ± 0%     18.7k ± 0%     ~     (all equal)
BackendBlockTraceQL/count                            10.1M ± 0%      9.9M ± 0%   -1.62%  (p=0.008 n=5+5)
BackendBlockTraceQL/desc                             6.56M ± 0%     6.47M ± 0%   -1.39%  (p=0.008 n=5+5)
```

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: tempodb/encoding/vparquet2/block_traceql.go [hunk 1]
```diff
@@ -1874,3 +1868,25 @@ func newSpanAttr(name string) traceql.Attribute {
...
+	case 0:
+		return nil
+	case 1:
+		return iters[0]
+	default:
+		return parquetquery.NewUnionIterator(definitionLevel, iters, pred)
+	}
+}
+
+func orIfNeeded(preds []parquetquery.Predicate) parquetquery.Predicate {
```
Comment (joe-elliott) [inline tempodb/encoding/vparquet2/block_traceql.go:1883]: i was noticing that `{ span.foo = "bar" && span.foo = "baz" }` creates the following:

```
LeftJoinIterator: 4: attributeCollector{}
required: 
	SyncIterator: rs.list.element.ss.list.element.Spans.list.element.Attrs.list.element.Key : InstrumentedPredicate{8, StringInPredicate{foo, foo, }}
optional: 
	SyncIterator: rs.list.element.ss.list.element.Spans.list.element.Attrs.list.element.Value : InstrumentedPredicate{0, OrPredicate{StringInPredicate{bar, },StringInPredicate{bat, }}}
```

- for the key iterator it's kind of an edge case, but we could write guard code that prevented adding dupes
- for the val iterator i'm wondering if a StringInPredicate{bar, bat, } works better? this behavior is the same as in main, so might not be worth addressing. unsure if it's even possible.

File: tempodb/encoding/vparquet2/block_traceql.go [hunk 2]
```diff
@@ -1874,3 +1868,25 @@ func newSpanAttr(name string) traceql.Attribute {
 func newResAttr(name string) traceql.Attribute {
 	return traceql.NewScopedAttribute(traceql.AttributeScopeResource, false, name)
 }
+
+func unionIfNeeded(definitionLevel int, iters []parquetquery.Iterator, pred parquetquery.GroupPredicate) parquetquery.Iterator {
+	switch len(iters) {
```
Comment (joe-elliott) [inline tempodb/encoding/vparquet2/block_traceql.go:1873]: are there savings here? one less set of `parquetquery.RowNumber`s to increment and track? if so i'm a bit surprised we're not seeing better #s on benchmarks.

in my profiling we spend a lot of time testing and incrementing row numbers so i would have expected better returns here. can we do something similar with the `LeftJoinIterator`?

File: tempodb/encoding/vparquet2/block_traceql.go [hunk 3]
```diff
@@ -1757,19 +1753,19 @@ type traceCollector struct {
 
 var _ parquetquery.GroupPredicate = (*traceCollector)(nil)
 
+func newTraceCollector() *traceCollector {
```
Comment (joe-elliott) [inline tempodb/encoding/vparquet2/block_traceql.go:1756]: the comments in this PR have been really helpful for me to understand some of the more esoteric lines. maybe add a comment here about the role of the tracecollector in creating spansets?

File: tempodb/encoding/vparquet2/block_traceql.go [hunk 4]
```diff
@@ -1779,41 +1775,39 @@ func (c *traceCollector) KeepGroup(res *parquetquery.IteratorResult) bool {
...
+			finalSpanset.Spans = append(finalSpanset.Spans, span)
+		}
+	}
+
+	// loop over all spans and add the trace-level attributes
+	for k, v := range c.traceAttrs {
+		for _, s := range finalSpanset.Spans {
+			s := s.(*span)
+			if _, alreadyExists := s.attributes[k]; !alreadyExists {
+				s.attributes[k] = v
```
Comment (joe-elliott) [inline tempodb/encoding/vparquet2/block_traceql.go:1809]: we do a lot of copying values onto the span struct into the one attributes map. i'm wondering if the span struct could contain multiple attribute maps for span/resource/trace/etc. levels. then we wouldn't need to copy the data. we could pass the same resource map to all attributes in the same batch.

the interface would have to change from this
```
func (s *span) Attributes() map[traceql.Attribute]traceql.Static {
	return s.attributes
}
```
to something like:
```
func (s *span) Attributes(traceqlAttribute) traceql.Static {
	 // check each map in precedence order
}
```

just an idle thought i had looking through the PR. not something for this PR, but do you think that would work?

**Review by joe-elliott — COMMENTED** [review]
nice change. i think this clarifies some questions i even had in the collectors. are there any important differences between vp2 and vp3? i didn't see any.

a few discussion points, but i'm good on approving.

**Comment by mdisibio** [comment]
Switched back to type checks instead of strings in res.OtherEntries and benchmarks are more clearcut as expected.  I had this originally but went to strings because it seemed more "correct" but micro-benchmarks showed that type checks are about significantly faster.  Should probably do the same for the remaining key checks but don't want to increase the scope of this PR any further.

```
name                                              old time/op    new time/op     delta
BackendBlockTraceQL/spanAttNameNoMatch              3.44ms ± 2%     3.44ms ± 2%     ~     (p=0.841 n=5+5)
BackendBlockTraceQL/spanAttValNoMatch               33.9ms ± 1%     33.6ms ± 0%     ~     (p=0.056 n=5+5)
BackendBlockTraceQL/spanAttValMatch                 38.3ms ± 1%     36.7ms ± 2%   -4.25%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicNoMatch         3.46ms ± 3%     3.46ms ± 1%     ~     (p=0.690 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicMatch           68.0ms ± 1%     61.7ms ± 1%   -9.32%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicRegexNoMatch    3.85ms ± 2%     3.80ms ± 0%     ~     (p=0.190 n=5+4)
BackendBlockTraceQL/spanAttIntrinsicRegexMatch       104ms ± 1%       97ms ± 1%   -6.38%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttNameNoMatch          3.06ms ± 2%     3.07ms ± 3%     ~     (p=0.548 n=5+5)
BackendBlockTraceQL/resourceAttValNoMatch           2.94ms ± 1%     2.95ms ± 2%     ~     (p=1.000 n=5+5)
BackendBlockTraceQL/resourceAttValMatch              606ms ± 2%      588ms ± 1%   -3.00%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch     2.97ms ± 2%     2.94ms ± 1%     ~     (p=0.421 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicMatch       63.8ms ± 1%     62.8ms ± 2%     ~     (p=0.056 n=5+5)
BackendBlockTraceQL/mixedNameNoMatch                 689ms ± 1%      695ms ± 1%     ~     (p=0.095 n=5+5)
BackendBlockTraceQL/mixedValNoMatch                  665ms ± 1%      663ms ± 1%     ~     (p=0.421 n=5+5)
BackendBlockTraceQL/mixedValMixedMatchAnd           3.11ms ± 2%     3.06ms ± 1%     ~     (p=0.222 n=5+5)
BackendBlockTraceQL/mixedValMixedMatchOr             556ms ± 1%      552ms ± 1%     ~     (p=0.095 n=5+5)
BackendBlockTraceQL/mixedValBothMatch               2.95ms ± 1%     2.97ms ± 3%     ~     (p=1.000 n=5+5)
BackendBlockTraceQL/count                            2.67s ± 4%      2.44s ± 3%   -8.84%  (p=0.008 n=5+5)
BackendBlockTraceQL/desc                             3.30s ± 1%      3.23s ± 1%   -2.05%  (p=0.008 n=5+5)

name                                              old speed      new speed       delta
BackendBlockTraceQL/spanAttNameNoMatch             932MB/s ± 2%    933MB/s ± 2%     ~     (p=0.841 n=5+5)
BackendBlockTraceQL/spanAttValNoMatch              228MB/s ± 1%    230MB/s ± 0%     ~     (p=0.056 n=5+5)
BackendBlockTraceQL/spanAttValMatch               97.4MB/s ± 1%  101.8MB/s ± 2%   +4.45%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicNoMatch        847MB/s ± 3%    847MB/s ± 1%     ~     (p=0.690 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicMatch          327MB/s ± 1%    361MB/s ± 1%  +10.28%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicRegexNoMatch   761MB/s ± 2%    769MB/s ± 0%     ~     (p=0.190 n=5+4)
BackendBlockTraceQL/spanAttIntrinsicRegexMatch     214MB/s ± 1%    229MB/s ± 1%   +6.82%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttNameNoMatch         412MB/s ± 2%    411MB/s ± 3%     ~     (p=0.548 n=5+5)
BackendBlockTraceQL/resourceAttValNoMatch          181MB/s ± 1%    180MB/s ± 2%     ~     (p=1.000 n=5+5)
BackendBlockTraceQL/resourceAttValMatch           64.4MB/s ± 2%   66.4MB/s ± 1%   +3.09%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch    226MB/s ± 2%    228MB/s ± 1%     ~     (p=0.421 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicMatch      572MB/s ± 1%    581MB/s ± 2%     ~     (p=0.056 n=5+5)
BackendBlockTraceQL/mixedNameNoMatch              19.0MB/s ± 1%   18.8MB/s ± 1%     ~     (p=0.079 n=5+5)
BackendBlockTraceQL/mixedValNoMatch               19.6MB/s ± 1%   19.7MB/s ± 1%     ~     (p=0.460 n=5+5)
BackendBlockTraceQL/mixedValMixedMatchAnd          406MB/s ± 2%    412MB/s ± 1%     ~     (p=0.222 n=5+5)
BackendBlockTraceQL/mixedValMixedMatchOr          45.9MB/s ± 1%   46.2MB/s ± 1%     ~     (p=0.087 n=5+5)
BackendBlockTraceQL/mixedValBothMatch              228MB/s ± 1%    226MB/s ± 3%     ~     (p=1.000 n=5+5)
BackendBlockTraceQL/count                         13.4MB/s ± 4%   14.7MB/s ± 3%   +9.68%  (p=0.008 n=5+5)
BackendBlockTraceQL/desc                          14.5MB/s ± 1%   14.8MB/s ± 1%   +2.09%  (p=0.008 n=5+5)

name                                              old MB_io/op   new MB_io/op    delta
BackendBlockTraceQL/spanAttNameNoMatch                3.21 ± 0%       3.21 ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttValNoMatch                 7.71 ± 0%       7.71 ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttValMatch                   3.73 ± 0%       3.73 ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicNoMatch           2.93 ± 0%       2.93 ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicMatch             22.3 ± 0%       22.3 ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicRegexNoMatch      2.93 ± 0%       2.93 ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicRegexMatch        22.3 ± 0%       22.3 ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttNameNoMatch            1.26 ± 0%       1.26 ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttValNoMatch             0.53 ± 0%       0.53 ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttValMatch               39.0 ± 0%       39.0 ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch       0.67 ± 0%       0.67 ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttIntrinsicMatch         36.5 ± 0%       36.5 ± 0%     ~     (all equal)
BackendBlockTraceQL/mixedNameNoMatch                  13.1 ± 0%       13.1 ± 0%     ~     (all equal)
BackendBlockTraceQL/mixedValNoMatch                   13.1 ± 0%       13.1 ± 0%     ~     (all equal)
BackendBlockTraceQL/mixedValMixedMatchAnd             1.26 ± 0%       1.26 ± 0%     ~     (all equal)
BackendBlockTraceQL/mixedValMixedMatchOr              25.5 ± 0%       25.5 ± 0%     ~     (all equal)
BackendBlockTraceQL/mixedValBothMatch                 0.67 ± 0%       0.67 ± 0%     ~     (all equal)
BackendBlockTraceQL/count                             35.7 ± 0%       35.7 ± 0%     ~     (all equal)
BackendBlockTraceQL/desc                              47.7 ± 0%       47.7 ± 0%     ~     (all equal)

name                                              old alloc/op   new alloc/op    delta
BackendBlockTraceQL/spanAttNameNoMatch              1.37MB ± 0%     1.37MB ± 0%   +0.00%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttValNoMatch               92.0MB ± 1%     91.7MB ± 1%     ~     (p=0.548 n=5+5)
BackendBlockTraceQL/spanAttValMatch                 1.38MB ± 0%     1.38MB ± 0%     ~     (p=0.310 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicNoMatch         1.59MB ± 0%     1.59MB ± 0%   +0.00%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicMatch           3.83MB ± 0%     3.86MB ± 0%   +0.60%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicRegexNoMatch    1.80MB ± 0%     1.80MB ± 0%   +0.00%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicRegexMatch      7.94MB ± 0%     7.96MB ± 0%   +0.25%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttNameNoMatch          1.27MB ± 0%     1.27MB ± 0%   +0.00%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttValNoMatch           1.27MB ± 0%     1.27MB ± 0%   +0.00%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttValMatch              184MB ± 0%      182MB ± 0%   -0.95%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch     1.27MB ± 0%     1.27MB ± 0%   +0.00%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicMatch       2.20MB ± 0%     2.19MB ± 0%   -0.39%  (p=0.008 n=5+5)
BackendBlockTraceQL/mixedNameNoMatch                 100MB ± 3%      103MB ± 0%     ~     (p=0.683 n=5+4)
BackendBlockTraceQL/mixedValNoMatch                  110MB ± 0%      109MB ± 1%     ~     (p=0.190 n=5+4)
BackendBlockTraceQL/mixedValMixedMatchAnd           1.27MB ± 0%     1.27MB ± 0%     ~     (p=0.151 n=5+5)
BackendBlockTraceQL/mixedValMixedMatchOr            3.95MB ± 0%     3.96MB ± 0%     ~     (p=0.190 n=5+4)
BackendBlockTraceQL/mixedValBothMatch               1.27MB ± 0%     1.27MB ± 0%   +0.00%  (p=0.016 n=5+5)
BackendBlockTraceQL/count                            936MB ± 0%      898MB ± 1%   -4.08%  (p=0.008 n=5+5)
BackendBlockTraceQL/desc                             690MB ± 0%      674MB ± 0%   -2.34%  (p=0.008 n=5+5)

name                                              old allocs/op  new allocs/op   delta
BackendBlockTraceQL/spanAttNameNoMatch               18.7k ± 0%      18.7k ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttValNoMatch                19.0k ± 0%      19.0k ± 0%     ~     (p=0.635 n=5+5)
BackendBlockTraceQL/spanAttValMatch                  19.1k ± 0%      19.1k ± 0%     ~     (p=0.333 n=4+5)
BackendBlockTraceQL/spanAttIntrinsicNoMatch          18.7k ± 0%      18.7k ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicMatch            30.1k ± 0%      30.1k ± 0%   -0.30%  (p=0.008 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicRegexNoMatch     21.0k ± 0%      21.0k ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicRegexMatch       73.0k ± 0%      72.9k ± 0%   -0.14%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttNameNoMatch           18.7k ± 0%      18.7k ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttValNoMatch            18.7k ± 0%      18.7k ± 0%     ~     (p=1.000 n=5+5)
BackendBlockTraceQL/resourceAttValMatch              1.69M ± 0%      1.68M ± 0%   -0.77%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch      18.7k ± 0%      18.7k ± 0%   +0.01%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicMatch        24.0k ± 0%      24.0k ± 0%   -0.11%  (p=0.016 n=5+4)
BackendBlockTraceQL/mixedNameNoMatch                 21.1k ± 0%      21.0k ± 0%   -0.05%  (p=0.008 n=5+5)
BackendBlockTraceQL/mixedValNoMatch                  21.9k ± 0%      21.9k ± 0%   -0.10%  (p=0.008 n=5+5)
BackendBlockTraceQL/mixedValMixedMatchAnd            18.7k ± 0%      18.7k ± 0%   -0.01%  (p=0.008 n=5+5)
BackendBlockTraceQL/mixedValMixedMatchOr             31.2k ± 0%      31.1k ± 0%   -0.35%  (p=0.008 n=5+5)
BackendBlockTraceQL/mixedValBothMatch                18.7k ± 0%      18.7k ± 0%     ~     (all equal)
BackendBlockTraceQL/count                            10.1M ± 0%       9.9M ± 0%   -1.62%  (p=0.008 n=5+5)
BackendBlockTraceQL/desc                             6.56M ± 0%      6.47M ± 0%   -1.39%  (p=0.008 n=5+5)
```

File: tempodb/encoding/vparquet2/block_traceql.go (hunk 1 above)
Comment (mdisibio) [reply tempodb/encoding/vparquet2/block_traceql.go:1877]: Yep this idea came up in the original vParquet and most of our predicates are designed for batches (RegexIn, StringIn).  They should be faster than the Or, but this type of optimization just hasn't made it to the top of the list.

**Comment by mdisibio** [comment]
> are there any important differences between vp2 and vp3?

No this part of the code should be identical between them.

File: tempodb/encoding/vparquet2/block_traceql.go (hunk 4 above)
Comment (mdisibio) [reply tempodb/encoding/vparquet2/block_traceql.go:1803]: Yes that sounds like a neat improvement.  Could do the same for future things like Scopes.

File: tempodb/encoding/vparquet2/block_traceql.go (hunk 2 above)
Comment (mdisibio) [reply tempodb/encoding/vparquet2/block_traceql.go:1867]: Right, less looping/peeking/copying of the intermediate results.  Savings depend on the query.  The only benchmark that triggers this is the new `{} | count() > 1` which is how I caught it but haven't measured the amount of improvement due to `unionIfNeeded` versus the batchCollector changes. 

It's hard to reason about but `{ span.http.status_code=500} | count() > 2` also triggers this and shows a nice improvement:

```
name                          old time/op    new time/op    delta
BackendBlockTraceQL/5xxcount    77.5ms ± 2%    70.2ms ± 1%   -9.40%  (p=0.000 n=10+9)

name                          old speed      new speed      delta
BackendBlockTraceQL/5xxcount   451MB/s ± 2%   498MB/s ± 1%  +10.36%  (p=0.000 n=10+9)

name                          old MB_io/op   new MB_io/op   delta
BackendBlockTraceQL/5xxcount      34.9 ± 0%      34.9 ± 0%     ~     (all equal)

name                          old alloc/op   new alloc/op   delta
BackendBlockTraceQL/5xxcount    1.78MB ± 0%    1.79MB ± 0%   +0.33%  (p=0.000 n=9+10)

name                          old allocs/op  new allocs/op  delta
BackendBlockTraceQL/5xxcount     22.8k ± 0%     22.7k ± 0%   -0.23%  (p=0.000 n=7+10)
```

**Review by joe-elliott — APPROVED** [review]

**Pull request merged**